mythril_core/src/ioapic.rs

```diff
@@ -0,0 +1,677 @@
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum DestinationMode {
+    /// Indicates that the destination is an APIC ID.
+    Physical = 0x00,
+    /// Indicates that the destination is a logical address.
+    Logical = 0x01,
+}
+
+impl TryFrom<u8> for DestinationMode {
+    type Error = Error;
+
+    fn try_from(value: u8) -> Result<DestinationMode> {
+        match value {
+            0x00 => Ok(DestinationMode::Physical),
+            0x01 => Ok(DestinationMode::Logical),
+            _ => Err(Error::InvalidValue(format!(
+                "Invalid destination mode: 0x{:x}",
+                value
+            ))),
+        }
+    }
+}
+
+/// Type of signal on the interrupt pin that triggers an interrupt.
+#[repr(u8)]
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum TriggerMode {
+    /// Edge sensitive trigger.
+    Edge = 0x00,
+    /// Level sensitive trigger.
+    Level = 0x01,
+}
+
+impl TryFrom<u8> for TriggerMode {
+    type Error = Error;
+
+    fn try_from(value: u8) -> Result<TriggerMode> {
+        match value {
+            0x00 => Ok(TriggerMode::Edge),
+            0x01 => Ok(TriggerMode::Level),
+            _ => Err(Error::InvalidValue(format!(
+                "Invalid trigger mode: 0x{:x}",
+                value
+            ))),
+        }
+    }
+}
+
+/// Polarity of the input signal.
+#[repr(u8)]
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum PinPolarity {
+    /// Active High
+    ActiveHigh = 0x00,
+    /// Active Low
+    ActiveLow = 0x01,
+}
+
+impl TryFrom<u8> for PinPolarity {
+    type Error = Error;
+
+    fn try_from(value: u8) -> Result<PinPolarity> {
+        match value {
+            0x00 => Ok(PinPolarity::ActiveHigh),
+            0x01 => Ok(PinPolarity::ActiveLow),
+            _ => Err(Error::InvalidValue(format!(
+                "Invalid pin polarity: 0x{:x}",
+                value
+            ))),
+        }
+    }
+}
+
+/// Delivery Mode
+///
+/// A 3-bit field that specifying how the destination APICs should act
+/// upon receiving the signal.
+#[repr(u8)]
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum DeliveryMode {
+    /// Deliver the signal on the INTR signal of all processor cores
+    /// listed in the destination.
+    Fixed = 0b000,
+    /// Deliver the signal on the INTR signal of all processor core
+    /// that is executing the lowest priority among listed processors.
+    LowestPriority = 0b001,
+    /// System Management Interrupt. Must be edge triggered and vector
+    /// must be all zeros.
+    SMI = 0b010,
+    /// Deliver the signal on the NMI signal of all processor cores
+    /// listed. Must be edge triggered.
+    NMI = 0b100,
+    /// All addressed local APICs will assume their INIT state. Must
+    /// be edge triggered.
+    INIT = 0b101,
+    /// Deliver the signal to the INTR signal of all processor cores
+    /// listed in destination. Must be edge triggered.
+    ExtINT = 0b111,
+}
+
+impl DeliveryMode {
+    fn level_is_valid(&self) -> bool {
```

Comment:
Maybe `valid_for_level_trigger`? Or something else, the current name seems confusing to me.